# this example is from GAPIT_with_SYslides.pdf
# the data is downloaded from the triticeatoolbox.org
# and has been filtered for 20% missing data and 5% > MAF (see below)

library(multtest)
library(gplots)
library(LDheatmap)
library(genetics)
library(ape)
library(EMMREML)
library(compiler) #this library is already installed in R
library("scatterplot3d")

source("http://zzlab.net/GAPIT/gapit_functions.txt")

source("http://zzlab.net/GAPIT/emma.txt")

# the actual data is in ./download_WZCG/
setwd('./wheatProtein/')

# traits
myY <- read.table("./download_WZCG/traits.txt", head = TRUE)

# genotypes as hapmap
# the tutorial uses read.table, but for TASSEL hapmat it is as follows
myG <- read.delim("./download_WZCG/genotype.hmp.txt", head = FALSE)

# quick poke around the data
str(myY)

## NOTE "778 obs", this is used below ##
##'data.frame':	778 obs. of  2 variables:
## $ X.Trait.              : Factor w/ 778 levels "07WA-601.22",..: 446 651 653 352 353 354 449 445 355 447 ...
## $ breeders_grain_protein: num  11.7 12.1 11.8 12.3 11.5 12.3 11.7 11.3 10.9 12.6 ...

# histogram of the phenotype data, bit of spread but with >700 lines
# expect >30 to be 3sds away from the mean
png('protein_histogram.png')
hist(myY$breeders_grain_protein)
dev.off()

# summary stats
mean(myY$breeders_grain_protein)
##[1] 12.04537
range(myY$breeders_grain_protein)
##[1]  9.6 16.7
sd(myY$breeders_grain_protein)
##[1] 1.01694

# look for lines missing data, this was pre-filtered, so is a good check!
which(is.na(myY$breeders_grain_protein))
##integer(0) # the expected result

# The tutorial will now analyse the data accounting for population
# structure with and without compression, using P3D
# When using hapmap format, GAPIT can impute missing data using the following:
## Major allele (SNP.impute="Major")
## Minor allele (SNP.impute="Minor")
## Heterozygote (SNP.impute="Middle")

# set output to new directory
setwd("./analysis1")
#Analysis1<-GAPIT(
#	Y = myY, 
#	G = myG,
#	SNP.impute="Major", # method of imputation
#	PCA.total=3, # PC for control population structure
#	Major.allele.zero=T,! # report allele effect with respect to minor alleel
#	group.from= 778, # see the results of str() command above
#	group.to=778,
#	group.by=1)

# the last three lines turn compression OFF because all the data is 
# used. # my understanding of compression is that you cluster data then
# use a central point to repressent all those data points. This reduces
# the possible model space.

# The Major.allele.zero controls the direction (sign) of the reported
# effect on the trait. For example + sign means, relative to the minor 
# effect allele, the allele is having a postive effect on the trait, and
# visa versa.

Analysis1<-GAPIT(
	Y = myY, 
	G = myG,
	SNP.impute="Major", 
	PCA.total=3, 
	Major.allele.zero=T,
	group.from=778,
	group.to=778,
	group.by=1)

# this next analysis uses a compression model
setwd('../analysis2')
Analysis2<-GAPIT(
	Y=myY,
	G=myG,
	SNP.impute="Major",
	PCA.total = 3,
	Major.allele.zero=T)

## comments on results
## PCA plots
# interestingly you get division between 2 row and 6 row barly, ie 
# real population structure is identified!! This split is along PC1
# as you would expect

## QQ plots
# analysis 1 get some deviation from expected, could be sign of issue.
# analysis 2 similar, lets try one of the other methods BIC
setwd('../BIC/')

BIC<-GAPIT(
	Y=myY,
	G=myG,
	SNP.impute="Major",
	PCA.total = 3,
	Major.allele.zero=T,
	Model.selection=T)

setwd('../SUPER/')
# try super model, needs a file containing covariente variable
# the has a header line that is Taxa PC1 PC2 PC3
# I am using the one previously generated by the other analysis above
# I am not sure if this is correct.
myCV<-read.csv('../analysis1/GAPIT.PCA.csv')
SUPER <- GAPIT(
	Y=myY[,c(1,2)],
	G=myG,
	CV=myCV,
	sangwich.top="MLM", 
	sangwich.bottom="SUPER", 
	LD=0.1,
	)
